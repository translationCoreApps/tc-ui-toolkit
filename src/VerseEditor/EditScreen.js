import React from 'react';
import PropTypes from 'prop-types';
import { getFontClassName } from '../common/fontUtils';
import { moveCursorToEnd } from './helpers/editHelpers';

/**
 * @callback EditScreen~onChange
 * @param {string} newVerse - the edited verse
 */

/**
 * Renders a text area for editing the verse
 * @property {string} verseText - the verse text to edit
 * @property {EditScreen~onChange} onChange - callback when the text has changed
 */
class EditScreen extends React.Component {
  constructor(props) {
    super(props);
    this._handleChange = this._handleChange.bind(this);
  }

  _handleChange(event) {
    const { onChange } = this.props;
    onChange(event.target.value);
  }

  render() {
    const {
      rows,
      style,
      verseText,
      targetLanguageFont,
    } = this.props;
    const fontClass = getFontClassName(targetLanguageFont);
    const className = fontClass ? `edit-screen ${fontClass}-text` : 'edit-screen';

    return (
      <textarea
        id="verse-editor-field"
        rows={rows}
        className={className}
        autoFocus={true}
        onFocus={moveCursorToEnd}
        onChange={this._handleChange}
        value={verseText}
        style={style}
      />
    );
  }
}

EditScreen.propTypes = {
  rows: PropTypes.number.isRequired,
  style: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  verseText: PropTypes.string.isRequired,
  targetLanguageFont: PropTypes.string,
};

EditScreen.defaultProps = {
  rows: 4,
  style: {},
};

export default EditScreen;
