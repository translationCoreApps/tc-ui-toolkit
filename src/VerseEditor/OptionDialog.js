import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import Toolbar from '@material-ui/core/Toolbar';
import { Glyphicon } from 'react-bootstrap';

const OptionDialog = ({
  isOpen,
  primaryOnclick,
  content,
  handleClose,
  headerTitleText,
  primaryButtonText,
  secondaryButtonText,
}) => {
  const actions = [
    <button
      key={1}
      className="btn-second"
      onClick={handleClose}
    >
      {secondaryButtonText}
    </button>,
    <button
      key={2}
      className="btn-prime"
      onClick={primaryOnclick}
    >
      {primaryButtonText}
    </button>,
  ];

  const headerContent = (
    <div style={{
      display: 'flex', justifyContent: 'space-between', width:'100%', marginLeft:20, marginRight:20,
    }}>
      <span style={{ color: 'var(--reverse-color)' }}>{headerTitleText}</span>
      <Glyphicon
        onClick={handleClose}
        glyph={'remove'}
        style={{
          color: 'var(--reverse-color)', cursor: 'pointer', fontSize: '18px', float: 'right',
        }}
      />
    </div>
  );

  return (
    <div>
      <Dialog
        open={isOpen}
        fullWidth
        onClose={handleClose}>
        <Toolbar disableGutters={true} style={{ backgroundColor: 'var(--accent-color-dark)' }}>
          {headerContent}
        </Toolbar>
        <br />
        <DialogContent style={{ padding: '0 18px 18px' }}>
          {content}
        </DialogContent>
        <DialogActions disableActionSpacing={true}>
          {actions}
        </DialogActions>
      </Dialog>
    </div>
  );
};

OptionDialog.propTypes = {
  handleClose: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  primaryOnclick: PropTypes.func.isRequired,
  content: PropTypes.any,
  headerTitleText: PropTypes.string.isRequired,
  primaryButtonText: PropTypes.string.isRequired,
  secondaryButtonText: PropTypes.string.isRequired,
};

export default OptionDialog;
