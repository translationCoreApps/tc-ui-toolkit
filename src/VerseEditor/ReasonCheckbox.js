import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from 'material-ui/Checkbox';

const styles = {
  root: {
    marginBottom: 16
  }
};

/**
 * @callback ReasonCheckbox~onCheck
 * @param {string} reason - the reason being checked
 * @param {bool} checked - indicates if the reason is checked
 */

/**
 * Renders a reason checkbox
 * @property {string} reason - the reason for the edit
 * @property {string} label - the checkbox label
 * @property {string[]} selectedReasons - an array of selected reasons
 * @property {ReasonCheckbox~onCheck} onCheck - callback when the checkbox is changed
 */
class ReasonCheckbox extends React.Component {

  constructor(props) {
    super(props);
    this._handleCheck = this._handleCheck.bind(this);
  }

  _handleCheck(e, checked) {
    const {reason, onCheck} = this.props;
    onCheck(reason, checked);
  }

  render() {
    const {reason, label, selectedReasons} = this.props;
    return (
      <Checkbox style={styles.root}
                checked={selectedReasons.includes(reason)}
                onCheck={this._handleCheck}
                label={label}/>
    );
  }
}

ReasonCheckbox.propTypes = {
  reason: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  selectedReasons: PropTypes.arrayOf(PropTypes.string),
  onCheck: PropTypes.func.isRequired
};

ReasonCheckbox.defaultProps = {
  selectedReasons: []
};

export default ReasonCheckbox;
