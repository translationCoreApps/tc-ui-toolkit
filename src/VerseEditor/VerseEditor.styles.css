@import '../index.styles.css';

svg > text {
  font-size: 1rem !important;
}

.edit-icon {
  color: #ffffff;
  width: 25;
  height: 25;
  margin-right: 5;
  margin-bottom: 5;
  vertical-align: middle;
}

.stepper-body {
  padding: 0;
}

.screen {
  padding: 24px;
}

.stepper {
  border-bottom: solid 1px #999;
  height: 50px;
}

.actions {
  padding: 0 24px;
  display: flex;
  justify-content: flex-end;
}

.done-icon {
  color: #ffffff;
  width: 20;
  height: 20;
  margin-right: 5;
  margin-bottom: 5;
  vertical-align: middle;
}

.reasons-screen {
  display: flex;
  justify-content: center;
}

.reasons-screen-column {
  display: flex;
  flex-direction: column;
}

.edit-screen {
  width: 100%;
  resize: none;
  padding: 10px;
  border: solid 1px var(--border-color);
  font-style: inherit;
  font-variant: inherit;
  font-weight: inherit;
  font-stretch: inherit;
  font-size: inherit;
  line-height: inherit;
  font-family: inherit;
  cursor: inherit;
  outline: none;
  background-color: transparent;
  -webkit-appearance: textfield;
  color: rgba(0, 0, 0, 0.870588);
}