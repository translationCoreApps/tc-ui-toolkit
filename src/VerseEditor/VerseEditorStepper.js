import React from 'react';
import PropTypes from 'prop-types';
import Stepper, {Step, StepLabel} from 'material-ui/Stepper';
import {withStyles} from 'material-ui/styles';

const styles = theme => ({
  label: {
    fontSize: 14
  }
});

/**
 * Renders the steps for editing a verse
 * @property {object} [style] - custom style attributes
 * @property {int} stepIndex - the current step index
 * @property {string[]} steps - an array of steps
 */
class VerseEditorStepper extends React.Component {
  render() {
    const {stepIndex, steps, style, classes} = this.props;
    return (
      <Stepper activeStep={stepIndex} style={style}>
        {steps.map((step, index) => {
          return (
            <Step key={index}>
              <StepLabel classes={{
                label: classes.label
              }}>{step}</StepLabel>
            </Step>
          );
        })}
      </Stepper>
    );
  }
}

VerseEditorStepper.propTypes = {
  style: PropTypes.object,
  stepIndex: PropTypes.number.isRequired,
  steps: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default withStyles(styles)(VerseEditorStepper);
