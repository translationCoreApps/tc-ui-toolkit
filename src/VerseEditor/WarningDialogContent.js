import React from 'react';
import PropTypes from 'prop-types';
import Logo from '../SpinningLogo/Logo';

const WarningDialogContent = ({ translate }) => (
  <div style={{ display: 'flex', alignItems: 'center' }}>
    <Logo
      height="100px"
      width="150px"
      style={{ margin: '15px 18px' }}
    />
    <p style={{ margin: '0 0 0 10px' }}>
      {translate('unsaved_changes')}
    </p>
  </div>
);

WarningDialogContent.propTypes = { translate: PropTypes.func.isRequired };

export default WarningDialogContent;
