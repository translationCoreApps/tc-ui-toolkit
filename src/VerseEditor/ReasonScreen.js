/* eslint-disable react/jsx-key */
import React from 'react';
import PropTypes from 'prop-types';
import ReasonCheckbox from './ReasonCheckbox';

export const updateReasons = (reason, checked, selectedReasons) => {
  let newReasons = [
    ...selectedReasons,
  ];

  if (checked && !newReasons.includes(reason)) {
    newReasons.push(reason);
  } else if (!checked && newReasons.includes(reason)) {
    const index = newReasons.indexOf(reason);
    newReasons.splice(index, 1);
  }
  return newReasons;
};

/**
 * @callback ReasonScreen~onChange
 * @param {string[]} newReasons - an array of reasons for editing the verse
 */

/**
 * Renders checkboxes to indicate the reason for the change
 * @property {ReasonScreen~onChange} onChange - callback when the selected reasons change
 * @property {func} translate - the locale function
 * @property {string[]} selectedReasons - an array of selected reasons
 */
class ReasonScreen extends React.Component {
  constructor(props) {
    super(props);
    this._handleCheck = this._handleCheck.bind(this);
  }

  /**
   * Checks if a checkbox is selected
   * @param {string} reason
   * @param {bool} checked
   * @return {boolean}
   * @private
   */
  _handleCheck(reason, checked) {
    const { selectedReasons, onChange } = this.props;
    const newReasons = updateReasons(reason, checked, selectedReasons);
    onChange(newReasons);
  }

  /**
   * get reasons as an array, so easy to separate into columns
   * @param {Function} translate
   * @param {Object} selectedReasons - selections state
   * @return {*[]}
   */
  getReasons(translate, selectedReasons) {
    const reasons = [
      <ReasonCheckbox reason="spelling"
        label={translate('spelling')}
        key={'spelling'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
      <ReasonCheckbox reason="punctuation"
        label={translate('punctuation')}
        key={'punctuation'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
      <ReasonCheckbox reason="word_choice"
        label={translate('word_choice')}
        key={'word_choice'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
      <ReasonCheckbox reason="meaning"
        label={translate('meaning')}
        key={'meaning'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
      <ReasonCheckbox reason="grammar"
        label={translate('grammar')}
        key={'grammar'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
      <ReasonCheckbox reason="other"
        label={translate('other')}
        key={'other'}
        onCheck={this._handleCheck}
        selectedReasons={selectedReasons}/>,
    ];
    return reasons;
  }

  /**
   * formats layout based on number of columns
   * @param {int} columns
   * @param {Array} reasons
   * @return {*}
   */
  getLayout(columns, reasons) {
    let perColumn = reasons.length / columns;
    const perColumnInt = Math.floor(perColumn);

    if (perColumnInt !== perColumn) {
      perColumn = perColumnInt + 1; // if not whole count
    }

    if (columns > 1) {
      return (
        <div className='reasons-screen'>
          {
            Array(columns).fill(0).map((value, index) => { // for each column
              const column = reasons.slice(index * perColumn, perColumn);
              return (
                <div className='reasons-screen-column'>
                  {column}
                </div>
              );
            })
          }
        </div>
      );
    }
    // else single column
    return (
      <div className='reasons-screen-column'>
        {reasons}
      </div>
    );
  }

  render() {
    const {
      translate, selectedReasons, columns,
    } = this.props;
    const reasons = this.getReasons(translate, selectedReasons);
    return this.getLayout(columns, reasons);
  }
}

ReasonScreen.propTypes = {
  onChange: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  selectedReasons: PropTypes.arrayOf(PropTypes.string).isRequired,
  columns: PropTypes.number.isRequired,
};

ReasonScreen.defaultProps = { columns: 2 };

export default ReasonScreen;
