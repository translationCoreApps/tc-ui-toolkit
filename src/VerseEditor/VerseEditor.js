import React from "react";
import PropTypes from 'prop-types';
import EditIcon from '@material-ui/icons/Edit';
import DoneIcon from '@material-ui/icons/Done';
// components
import EditScreen from './EditScreen';
import ReasonScreen from './ReasonScreen';
import BaseDialog from './BaseDialog';
import VerseEditorStepper from './VerseEditorStepper';

import './VerseEditor.styles.css';

const steps = ['edit_verse', 'select_reasons'];

/**
 * Checks if the next butt should be enabled
 * @param state
 * @return {*}
 */
export const isNextEnabled = (state) => {
  const {stepIndex, verseChanged, newVerse, reasons} = state;
  switch (stepIndex) {
    case 0:
      return verseChanged && Boolean(newVerse);
    case 1:
      return reasons.length > 0;
    default:
      return false;
  }
};

/**
 * @callback VerseEditor~submitCallback
 * @param {string} originalVerse - the original verse text
 * @param {string} newVerse - the edited version of the verse text
 * @param {string[]} reasons - an array of reasons for editing the verse
 */

/**
 * Renders a form for editing a single verse
 * @property {string} verseText - the verse text to edit
 * @property {func} translate - the locale function
 * @property {VerseEditor~submitCallback} onSubmit - callback when the edit is submitted
 * @property {func} onCancel - callback when the edit is canceled
 */
class VerseEditor extends React.Component {

  constructor(props) {
    super(props);
    this._handleBack = this._handleBack.bind(this);
    this._handleCancel = this._handleCancel.bind(this);
    this._handleNext = this._handleNext.bind(this);
    this._isLastStep = this._isLastStep.bind(this);
    this._handleVerseChange = this._handleVerseChange.bind(this);
    this._handleReasonChange = this._handleReasonChange.bind(this);
    this._resetState = this._resetState.bind(this);
    this.state = {
      stepIndex: 0,
      newVerse: '',
      verseChanged: false,
      reasons: []
    };
  }

  _resetState() {
    this.setState({
      stepIndex: 0,
      newVerse: '',
      verseChanged: false,
      reasons: []
    });
  }

  _handleBack() {
    const {stepIndex} = this.state;
    this.setState({
      stepIndex: Math.max(stepIndex - 1, 0)
    });
  }

  _handleCancel() {
    const {onCancel} = this.props;
    onCancel();
    this._resetState();
  }

  _handleNext() {
    const {stepIndex, newVerse, reasons} = this.state;
    const {verseText, onSubmit} = this.props;
    if (this._isLastStep()) {
      onSubmit(verseText, newVerse, reasons);
      this._resetState();
    } else {
      this.setState({
        stepIndex: stepIndex + 1
      });
    }
  }

  _handleVerseChange(newVerse) {
    const {verseText} = this.props;
    this.setState({
      newVerse: newVerse,
      verseChanged: newVerse !== verseText
    });
  }

  _handleReasonChange(newReasons) {
    this.setState({
      reasons: newReasons
    });
  }


  _isLastStep() {
    const {stepIndex} = this.state;
    return stepIndex === steps.length - 1;
  }

  /**
   * Checks if the next button is enabled
   * @return {*}
   */
  _isNextEnabled() {
    return isNextEnabled(this.state);
  }

  render() {
    const {translate, open, verseTitle, verseText} = this.props;
    const {stepIndex, newVerse, reasons} = this.state;
    let text = !this.state.verseChanged ? verseText : newVerse;
    let screen;
    switch (stepIndex) {
      case 0:
        screen = (<EditScreen verseText={text} onChange={this._handleVerseChange} />);
        break;
      case 1:
        screen = (<ReasonScreen translate={translate} selectedReasons={reasons} onChange={this._handleReasonChange} />);
        break;
      default:
        screen = translate('oops');
    }

    let nextStepButtonTitle = translate('buttons.next_button');
    if (this._isLastStep()) {
      nextStepButtonTitle = (
        <React.Fragment>
          <DoneIcon className='done-icon' />
          {translate('buttons.save_button')}
        </React.Fragment>
      );
    }

    const localizedSteps = [];
    for (const step of steps) {
      localizedSteps.push(translate(step));
    }

    const title = (
      <span>
        <EditIcon className='edit-icon' />
        {translate('edit_verse_title', {passage: verseTitle})}
      </span>
    );

    return (
      <BaseDialog modal={true}
        open={open}
        title={title}>
        <VerseEditorStepper stepIndex={stepIndex}
          className='stepper'
          steps={localizedSteps} />
        <div className='screen'>
          {screen}
        </div>
        <div className='actions'>
          <button className="btn btn-link"
            disabled={stepIndex === 0}
            style={{color: stepIndex === 0 ? '#777' : 'var(--accent-color-dark)'}}
            onClick={this._handleBack}>
            {translate('buttons.back_button')}
          </button>
          <button className="btn-second"
            onClick={this._handleCancel}>
            {translate('buttons.cancel_button')}
          </button>
          <button className="btn-prime"
            disabled={!this._isNextEnabled()}
            onClick={this._handleNext}>
            {nextStepButtonTitle}
          </button>
        </div>
      </BaseDialog>
    );
  }
}

VerseEditor.propTypes = {
  open: PropTypes.bool.isRequired,
  verseTitle: PropTypes.string.isRequired,
  verseText: PropTypes.string.isRequired,
  translate: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

export default VerseEditor;