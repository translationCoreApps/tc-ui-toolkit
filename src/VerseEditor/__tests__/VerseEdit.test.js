/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import {createMuiTheme, MuiThemeProvider} from '@material-ui/core/styles';
import VerseEditor from '../VerseEditor';

const mock_translate = (text) => (text);

describe('VerseEditor component:', () => {

  test('should render', () => {
    // given
    const theme = createMuiTheme({
      typography: { useNextVariants: true },
      scrollbarThumb: { borderRadius: '10px' },
    });
    const props = getBasePropertiesAndMockActions();

    // when
    const component = renderer.create(
      <MuiThemeProvider theme={theme}>
        <VerseEditor {...props} />
      </MuiThemeProvider>
    );

    // then
    expect(component.toJSON()).toMatchSnapshot();
  });
});

//
// helpers
//

function getBasePropertiesAndMockActions() {
  const props = {
    open: true,
    verseTitle: 'Dummy Title',
    verseText: 'Dummy Text',
    translate: mock_translate,
    onCancel: () => jest.fn(),
    onSubmit: () => jest.fn(),
    targetLanguage: 'Klingon',
  };

  return props;
}
