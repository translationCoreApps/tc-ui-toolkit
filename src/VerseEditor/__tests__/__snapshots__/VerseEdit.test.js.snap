// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerseEditor component: should render 1`] = `null`;
